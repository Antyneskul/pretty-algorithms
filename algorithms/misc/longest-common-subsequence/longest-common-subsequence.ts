import { fill, range } from '../../utils';

/**
 * The longest common subsequence (LCS) problem is the problem of finding the
 * longest subsequence common to two sequences.
 *
 * It differs from problems of finding common substrings since subsequences are
 * not required to occupy consecutive positions within the original sequences.
 */

export type Directions = Array<Array<'←' | '↑' | '↖'>>;

export interface BestResults {
  lengths: number[][];
  directions: Directions;
}

export function lcsLengths(seqA: string, seqB: string): BestResults {
  const lengthA = seqA.length;
  const lengthB = seqB.length;

  const lengths: number[][] = new Array(lengthA + 1);
  fill(lengths, () => new Array(lengthB + 1));

  const directions: Directions = new Array(lengthA + 1);
  fill(directions, () => new Array(lengthB + 1));

  range(0, lengthA).forEach(i => (lengths[i][0] = 0));
  range(0, lengthB).forEach(i => (lengths[0][i] = 0));

  range(0, lengthA - 1).forEach(indexA => {
    range(0, lengthB - 1).forEach(indexB => {
      const charA = seqA[indexA];
      const charB = seqB[indexB];

      if (charA === charB) {
        lengths[indexA + 1][indexB + 1] = lengths[indexA][indexB] + 1;
        directions[indexA + 1][indexB + 1] = '↖';
      } else if (lengths[indexA][indexB + 1] >= lengths[indexA + 1][indexB]) {
        lengths[indexA + 1][indexB + 1] = lengths[indexA][indexB + 1];
        directions[indexA + 1][indexB + 1] = '↑';
      } else {
        lengths[indexA + 1][indexB + 1] = lengths[indexA + 1][indexB];
        directions[indexA + 1][indexB + 1] = '←';
      }
    });
  });

  return {
    lengths,
    directions,
  };
}

function walkLCS(directions: Directions, seqA: string, indexA: number, indexB: number): string {
  if (indexA === 0 || indexB === 0) return '';

  const direction = directions[indexA][indexB];

  if (direction === '↖') {
    const subLCS = walkLCS(directions, seqA, indexA - 1, indexB - 1);
    return subLCS + seqA[indexA - 1];
  } else if (direction === '←') {
    return walkLCS(directions, seqA, indexA, indexB - 1);
  } else {
    return walkLCS(directions, seqA, indexA - 1, indexB);
  }
}

/**
 * Return the longest common subsequence (LCS)
 * @param seqA First sequence
 * @param seqB Second sequence
 */
export function findLCS(seqA: string, seqB: string): string {
  const { directions } = lcsLengths(seqA, seqB);
  const lcs = walkLCS(directions, seqA, seqA.length, seqB.length);

  return lcs;
}
