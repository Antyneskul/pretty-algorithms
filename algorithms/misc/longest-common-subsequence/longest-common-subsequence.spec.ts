import { lcsLengths, findLCS } from './longest-common-subsequence';

describe('lcsLengths', () => {
  it('should return the table of common subsequences lengths', () => {
    const seqA = 'ABCBDAB';
    const seqB = 'BDCABA';
    const { lengths } = lcsLengths(seqA, seqB);

    expect(lengths).toMatchSnapshot('LCS subsequences lengths');
  });

  it('should return the table of directions to recreate the LCS', () => {
    const seqA = 'ABCBDAB';
    const seqB = 'BDCABA';
    const { directions } = lcsLengths(seqA, seqB);

    expect(directions).toMatchSnapshot('LCS directions');
  });
});

describe('findLCS', () => {
  it('should return the LCS for two sequences', () => {
    const seqA = 'ABCBDAB';
    const seqB = 'BDCABA';
    const lcs = findLCS(seqA, seqB);

    expect(lcs).toBe('BCBA');
  });
});
