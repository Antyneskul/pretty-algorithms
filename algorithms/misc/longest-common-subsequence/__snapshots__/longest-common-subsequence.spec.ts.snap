// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LCS directions 1`] = `
Array [
  Array [
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
    undefined,
  ],
  Array [
    undefined,
    "↑",
    "↑",
    "↑",
    "↖",
    "←",
    "↖",
  ],
  Array [
    undefined,
    "↖",
    "←",
    "←",
    "↑",
    "↖",
    "←",
  ],
  Array [
    undefined,
    "↑",
    "↑",
    "↖",
    "←",
    "↑",
    "↑",
  ],
  Array [
    undefined,
    "↖",
    "↑",
    "↑",
    "↑",
    "↖",
    "←",
  ],
  Array [
    undefined,
    "↑",
    "↖",
    "↑",
    "↑",
    "↑",
    "↑",
  ],
  Array [
    undefined,
    "↑",
    "↑",
    "↑",
    "↖",
    "↑",
    "↖",
  ],
  Array [
    undefined,
    "↖",
    "↑",
    "↑",
    "↑",
    "↖",
    "↑",
  ],
]
`;

exports[`LCS subsequences lengths 1`] = `
Array [
  Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  Array [
    0,
    0,
    0,
    0,
    1,
    1,
    1,
  ],
  Array [
    0,
    1,
    1,
    1,
    1,
    2,
    2,
  ],
  Array [
    0,
    1,
    1,
    2,
    2,
    2,
    2,
  ],
  Array [
    0,
    1,
    1,
    2,
    2,
    3,
    3,
  ],
  Array [
    0,
    1,
    2,
    2,
    2,
    3,
    3,
  ],
  Array [
    0,
    1,
    2,
    2,
    3,
    3,
    4,
  ],
  Array [
    0,
    1,
    2,
    2,
    3,
    4,
    4,
  ],
]
`;
